runtime/lib/backend/openqasm/OpenQasmBuilder.hpp

```diff
@@ -0,0 +1,409 @@
+class QasmRegister {
+  private:
+    const RegisterType type;
+    const std::string name;
+    size_t size;
+
+  public:
+    explicit QasmRegister(RegisterType _type, const std::string &_name, size_t _size)
+        : type(_type), name(_name), size(_size)
+    {
+    }
+    ~QasmRegister() = default;
+
+    [[nodiscard]] auto getType() const -> RegisterType { return this->type; }
+    [[nodiscard]] auto getName() const -> std::string { return this->name; }
+    [[nodiscard]] auto getSize() const -> size_t { return this->size; }
+
+    void updateSize(size_t new_size) { this->size = new_size; }
+    void resetSize() { this->size = 0; }
+    auto isValidSlice(const std::vector<size_t> &slice) const -> bool
+    {
+        if (slice.empty()) {
+            return false;
+        }
+
+        return std::all_of(slice.begin(), slice.end(),
+                           [this](auto qubit) { return this->size > qubit; });
+    }
+
+    auto toOpenQasm(RegisterMode mode, [[maybe_unused]] const std::vector<size_t> &slice = {},
+                    [[maybe_unused]] const std::string &version = "3.0") const -> std::string
+    {
+        std::ostringstream oss;
+        switch (mode) {
+        case RegisterMode::Alloc: {
+            // qubit[size] name;
+            if (this->type == RegisterType::Qubit) {
+                oss << "qubit";
+            }
+            else if (this->type == RegisterType::Bit) {
+                oss << "bit";
+            }
+            else {
+                RT_FAIL("Unsupported OpenQasm register type");
+            }
+            oss << "[" << this->size << "] " << this->name << ";\n";
+            return oss.str();
+        }
+        case RegisterMode::Reset: {
+            // reset name;
+            oss << "reset " << this->name << ";\n";
+            return oss.str();
+        }
+        case RegisterMode::Slice: {
+            // name[slice_0], ..., name[slice_n]
+            RT_ASSERT(isValidSlice(slice));
+            auto iter = slice.begin();
+            for (; iter != slice.end() - 1; iter++) {
+                oss << this->name << "[" << *iter << "], ";
+            }
+            oss << this->name << "[" << *iter << "]";
+            return oss.str();
+        }
+        default:
+            RT_FAIL("Unsupported OpenQasm register mode");
+        }
+    }
+};
+
+/**
+ * The OpenQasm gate type.
+ *
+ * @param name The name of the gate to apply from the list of supported gates
+ * (`rt_qasm_gate_map`)
+ * @param params_val Optional list of parameter values for parametric gates
+ * @param params_str Optional list of parameter names for parametric gates
+ * @param wires Wires to apply gate to
+ * @param inverse Indicates whether to use inverse of gate
+ *
+ * @note Parametric gates are currently supported via either their values or names
+ * but not both.
+ */
+class QasmGate {
+  private:
+    const std::string name;
+    const std::vector<double> params_val;
+    const std::vector<std::string> params_str;
```

Comment:
The parametrize gates are not yet supported from the OQ device and this is part of the future work. However, the builder is offering a basic support for such gates. 

> Can a gate have mixed string/double parameters?

No, it's not possible at the moment. But, this will be revisited certainly in future for adding the param gates support. Trying the mix of params would raise a runtime error. 

> Is it correct that both val and str are possible values of gate parameters? Or, alternatively, does param_str means to contain variable names?

Yes. the `param_str` is an optional list of parameter names for parametric gates.